Mathlib/GroupTheory/Coxeter/Matrix.lean

```diff
@@ -29,48 +51,99 @@ This file defines some standard Coxeter matrices.
 * `CoxeterMatrix.H₄` : Coxeter matrix for the symmetry group of the regular 4-polytopes,
   the 120-cell and 600-cell.
 
+## References
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 4--5, 13--15
+
+* [J. Baez, *Coxeter and Dynkin Diagrams*](https://math.ucr.edu/home/baez/twf_dynkin.pdf)
+
 -/
 
+universe u
+
+/-- A *Coxeter matrix* is a symmetric matrix of natural numbers whose diagonal entries are equal to
+1 and whose off-diagonal entries are not equal to 1. -/
+def CoxeterMatrix (B : Type u) : Type u :=
+  {M : Matrix B B ℕ // (∀ i i', M i i' = M i' i) ∧ (∀ i, M i i = 1) ∧ ∀ i i', i ≠ i' → M i i' ≠ 1}
```

Comment:
What was the reason for not using `M.IsSymm` any more? Does #11878 resolve it?

Reply:
There are now two different constructors for `CoxeterMatrix`; first, `CoxeterMatrix.of`, which uses a `∀ i i', M i i' = M i' i`, and second, `CoxeterMatrix.of'`, which uses an `M.IsSymm`. Also, there is a theorem `CoxeterMatrix.symmetric` that shows `∀ i i', M i i' = M i' i` and another theorem `CoxeterMatrix.isSymm` that shows `M.IsSymm`. Because `∀ i i', M i i' = M i' i` or `M.IsSymm` are interchangeable in this way, I am not convinced that it matters that much which one is used internally as part of the definition.

Each way of writing the condition has its advantages, so I think it is good to supply both. For example, `M 2 3` can be rewritten to `M 3 2` using `M.symmetric`, which is easier than using `M.isSymm.apply`. But when we define the direct product of Coxeter groups, we will probably want to use [Matrix.IsSymm.fromBlocks](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Matrix/Symmetric.html#Matrix.IsSymm.fromBlocks) on the Coxeter matrices of each component, which requires a `Matrix.IsSymm`.

Reply:
Taking into account @eric-wieser remarks, I think the best solution would be:
```lean
structure CoxeterMatrix (B : Type u) where
  M : Matrix B B ℕ
  isSymm : M.IsSymm
  diagonal i : M i i = 1
  off_diagonal i i' : i ≠ i' → M i i' ≠ 1
```

Reply:
I still suspect that `IsCoxeter` has value; for one, it means you can write code like `if M.IsCoxeter then _ else _`, using #11879.

Reply:
I do not see any situation where we would like to write `if M.IsCoxether then _ else _`... If #11879 facilitates automation and if this requires a separate a structure, we may have such a structure, but then I would advocate for:
```lean
structure CoxeterMatrix (B : Type u) where
  M : Matrix B B ℕ
  isCoxeter : M.IsCoxeter
```
(Or a subtype, but then, I may prefer `def` over `abbrev`.)

In any case, what I feel is the most important, from a mathematical point of view, is that the API for Coxeter groups (e.g. the presented group, `CoxeterSystem`) take as an input something in `CoxeterMatrix B`, and *not* two separate data `M` and `isCoxeter`, as it was the case before this refactor PR.

Reply:
Hello, I changed it to a structure.

Reply:
I am running into the following problem now.

```
/-- A *Coxeter matrix* is a symmetric matrix of natural numbers whose diagonal entries are equal to
1 and whose off-diagonal entries are not equal to 1. -/
@[ext]
structure CoxeterMatrix (B : Type u) where
  M : Matrix B B ℕ
  isSymm : M.IsSymm := by decide
  diagonal i : M i i = 1 := by decide
  off_diagonal i i' : i ≠ i' → M i i' ≠ 1 := by decide

variable {B : Type*}

/-- A Coxeter matrix can be coerced to a matrix. -/
instance : Coe (CoxeterMatrix B) (Matrix B B ℕ) := ⟨M⟩

/-- A Coxeter matrix can be thought of as a function `B → B → ℕ`. -/
instance : FunLike (CoxeterMatrix B) B (B → ℕ) := ⟨M, CoxeterMatrix.ext⟩

attribute [simp] CoxeterMatrix.diagonal
```

I want `M i i` to simplify to 1 if `M` is a Coxeter matrix. But it won't, because the expression `M i i` uses the `FunLike` instance, which means it's not syntactically equal to `(M : Matrix B B ℕ) i i = M.M i i`, which means that the `[simp]` theorem `CoxeterMatrix.diagonal` won't apply to it. What is the best solution to this?

Reply:
I figured it out, I think.